compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td

```diff
@@ -678,6 +679,43 @@ def HAL_AllocatorImportOp : HAL_Op<"allocator.import", [
   }];
 }
 
+def HAL_AllocatorResolveMemoryPropertiesOp : HAL_PureOp<"allocator.resolve_memory_properties", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
+]> {
+  let summary = [{Resolves memory properties from resource lifetime and affinity.}];
+  let description = [{
+    Resolves the required memory types and buffer usage for a resource with
+    the given lifetime and affinity. This operation encapsulates the logic
+    for deriving buffer properties based on stream resource semantics.
+
+    This operation can be resolved later in compilation either by canonicalization
+    for single device affinties or using the topology attribute for multiple devices.
+
+    Example:
+    ```mlir
+    %memory_types, %buffer_usage = hal.allocator.resolve_memory_properties
+        for(#hal.device.affinity<@device_a>)
+        lifetime(transient) : i32, i32
+    ```
+  }];
+
+  let arguments = (ins
+    OptionalAttr<Stream_AffinityAttr>:$affinity,
```

Comment:
HAL cannot depend on stream (as more than just stream may lower into HAL) - remove the StreamBase.td include and change this to `OptionalAttr<AnyAttr>` (HAL doesn't care about what the affinity is, only that it can be resolved). We should probably have a HAL affinity attr interface but I'm not sure it's possible to external model an attr interface like it is op ones.